shards/modules/channels/channels.hpp

```diff
@@ -25,55 +24,61 @@ struct DummyChannel : public ChannelShared {
 };
 
 struct MPMCChannel : public ChannelShared {
-  MPMCChannel(bool noCopy) : ChannelShared(), _noCopy(noCopy) {}
-
-  // no real cleanups happens in Produce/Consume to keep things simple
-  // and without locks
-  virtual ~MPMCChannel() {
-    if (!_noCopy) {
-      SHVar tmp{};
-      while (data.pop(tmp)) {
-        destroyVar(tmp);
-      }
-      while (recycle.pop(tmp)) {
-        destroyVar(tmp);
-      }
-    }
-  }
+  MPMCChannel() : ChannelShared() {}
+
+  virtual ~MPMCChannel() {}
 
   virtual void clear() override {
-    SHVar tmp{};
-    while (data.pop(tmp)) {
-      destroyVar(tmp);
-    }
-    while (recycle.pop(tmp)) {
-      destroyVar(tmp);
+    // move all to recycle
+    OwnedVar value{};
+    while (_data.try_pop(value)) {
+      _recycle.push(std::move(value));
     }
   }
 
-  // A single source to steal data from
-  boost::lockfree::queue<SHVar> data{16};
-  boost::lockfree::stack<SHVar> recycle{16};
+  void push(OwnedVar &&value) { _data.push(std::move(value)); }
+
+  void recycle(OwnedVar &&value) { _recycle.push(std::move(value)); }
+
+  void push(const SHVar &value) {
+    // in this case try check recycle bin
+    OwnedVar valueClone{};
+    _recycle.try_pop(valueClone);
+    valueClone = value;
+    _data.push(std::move(valueClone));
+  }
+
+  bool try_unrecycle(OwnedVar &value) { return _recycle.try_pop(value); }
+
+  template <bool Recycle = true, typename Func> bool try_pop(Func &&func) {
+    OwnedVar value{};
+    if (_data.try_pop(value)) {
+      func(std::forward<OwnedVar>(value));
+      if constexpr (Recycle) {
+        _recycle.push(std::move(value));
```

Comment:
Maybe add a limit , cause if you only use the moving push it might grow the recycle list endlessly